feat: Email made translatable

codedsun opened the pull request:
<!--
(Thanks for sending a pull request! Please make sure you click the link above to view the contribution guidelines, then fill out the blanks below.)
-->
<!-- Add the issue number that is fixed by this PR (In the form Fixes #123) -->

Fixes #7405 

#### Short description of what this resolves:


#### Changes proposed in this pull request:

-
-
-

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes <!-- use `nosetests tests/` to run all the tests -->
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
<!-- If an existing function does not have a docstring, please add one -->
- [x] All the functions created/modified in this PR contain relevant docstrings.

iamareebjamal commented on app/api/auth.py:
```diff
@@ -342,19 +343,23 @@ def reset_password_post():
+            send_email(
+                to=user,
+                action=PASSWORD_RESET_AND_VERIFY,
+                html=render_template('email/password_reset_and_verify', **context),
```
pass context as keyword arguments

iamareebjamal replied:
.html missing

iamareebjamal commented on app/api/auth.py:
```diff
@@ -342,19 +343,23 @@ def reset_password_post():
+            send_email(
+                to=user,
+                action=PASSWORD_RESET,
+                html=render_template('email/password_reset', **context),
```
pass context as keyword arguments

iamareebjamal replied:
.html missing

iamareebjamal commented on app/api/users.py:
```diff
@@ -93,13 +94,17 @@ def after_create_object(self, user, data, view_kwargs):
+            to=user,
+            action=USER_REGISTER,
+            subject=MAILS[USER_REGISTER]['subject'].format(app_name=settings['app_name']),
+            html=render_template('email/user_register.html', **context),
```
pass context as keyword arguments

iamareebjamal commented on app/api/auth.py:
```diff
@@ -413,8 +418,14 @@ def change_password():
+                subject=MAILS[PASSWORD_CHANGE]['subject'].format(
+                    app_name=get_settings()['app_name']
+                ),
+                html=render_template('email/password_change'),
```
.html missing

iamareebjamal commented on app/templates/email/password_reset.html:
```diff
@@ -0,0 +1,3 @@
+Please use the following link to reset your password.
+<br/><a href='{link}' target='_blank'>{{ link }}</a>
```
```suggestion
<br/>{{ link }}
```

iamareebjamal commented on app/templates/email/password_reset_and_verify.html:
```diff
@@ -0,0 +1,2 @@
+Please use the following link to reset your password and verify your account.
+<br> <a href='{link}' target='_blank'> {{ link }} </a>
```
```suggestion
<br>{{ link }}
```

codedsun commented:
@iamareebjamal - mo file created.

iamareebjamal commented on app/templates/email/ticket_purchased.html:
```diff
@@ -1,19 +1,19 @@
+<br/><br/>{{ _('This is your order confirmation. You can download your tickets at') }} : {{ order_view_url }}
+<br/><br/>{{ _('Order Summary') }}:
 <br/>Order: {{ order.invoice_number }}
 <br/>Name: {{ order.user.full_name }}
```
_() missing here

iamareebjamal commented on app/templates/email/ticket_purchased_attendee.html:
```diff
@@ -1,17 +1,17 @@
+{{ _('Hello') }},
+<br/><br/>{{ _('This is a confirmation mail of your tickets for the event') }} {{ order.event.name }}. {{ _('You can download your tickets at') }}: {{ order_view_url }}
+<br/><br/>{{ _('Ticket Summary') }}:
 <br/>Name: {{ attendees.0.user.full_name }}
```
_() missing here

iamareebjamal commented on app/templates/email/user_register.html:
```diff
@@ -0,0 +1,7 @@
+{{ _('Hello') }},
+<br/><br/>{{ _('Your account has been created on') }} {{ settings.app_name }}. {{ _('Congratulations! ') }}
```
```suggestion
<br/><br/>{{ _('Your account has been created on') }} {{ settings.app_name }}. {{ _('Congratulations!') }}
```

iamareebjamal commented on app/templates/email/user_register.html:
```diff
@@ -0,0 +1,7 @@
+{{ _('Hello') }},
+<br/><br/>{{ _('Your account has been created on') }} {{ settings.app_name }}. {{ _('Congratulations! ') }}
+<br/><br/>{{ _('Your login is: {{ email }} ')}}
```
email should be out of _()

iamareebjamal commented on app/templates/email/user_register.html:
```diff
@@ -0,0 +1,7 @@
+<br/><br/>{{ _('Your login is: {{ email }} ')}}
+<br/><br/>{{ _('Please visit the following link to verify your email:') }} {{ link }}
+<br/><br/>{{ _('Thank You') }},
+<br/><br/>{{ settings.app_name }} Team
```
_() missing here

codedsun commented:
@iamareebjamal - While passing context as keyword argument in send_email function , the jinja html template doesn't recognize values and give error object not found.

iamareebjamal commented:
You don't pass context as kwarg, you pass contents of context as kwargs

iamareebjamal commented on app/api/users.py:
```diff
@@ -93,13 +94,17 @@ def after_create_object(self, user, data, view_kwargs):
+            action=USER_REGISTER,
+            subject=MAILS[USER_REGISTER]['subject'].format(app_name=settings['app_name']),
+            html=render_template('email/user_register.html', context=context),
+            attachments=None,
```
If attachments is None, then this is not needed

iamareebjamal commented on config.py:
```diff
@@ -49,6 +49,7 @@ class Config:
     FLASK_ADMIN_SWATCH = 'lumen'
 
     VERSION = VERSION_NAME
+    LANGUAGES = ['en', 'de', 'fr']
```
Add all languages present in the translations folder

iamareebjamal commented on babel.cfg:
```diff
@@ -0,0 +1,2 @@
+[jinja2: app/templates/**.html]
```
Add python files here as well

iamareebjamal commented on messages.pot:
```diff
@@ -0,0 +1,125 @@
+# Translations template for PROJECT.
```
Delete this file

iamareebjamal reviewed (changes requested):
Add required dependencies in requirements

codedsun commented:
@iamareebjamal - Done the changes

iamareebjamal commented on requirements/common.txt:
```diff
@@ -62,6 +62,7 @@ pyyaml==5.3.1
 sendgrid==6.4.7
 marshmallow==2.15.2
 WeasyPrint==52.1
+Flask-Babel=2.0.0
```
wrong syntax

iamareebjamal replied:
```suggestion
Flask-Babel==2.0.0
```

iamareebjamal commented:
Please add a screenshot of the mail you received and fix the build

codedsun commented:
@iamareebjamal For every mail we save the text in the `mails` table. Can  i show you what was present there under column message.

```
Hello,
<br/><br/>Sunet Srivastava Event Till March.
<br/><br/>This is your order confirmation. You can download your tickets at : http://eventyay.com/orders/edd5f2a3-c44e-44fd-905b-4019ad975b4f/view
<br/><br/>Order Summary:
<br/>Order: O1604747108-22
<br/>Name: User User
<br/>

    <br/>'Ticket': Free Ticket
    <br/> 'Suneet': 1
    <br/>

<br/>About this event :
<br/>Datum: October 31, 2020 02:10 PM - March 01, 2021 05:00 PM
<br/>Online
<br/><br/>Best regards,
<br/>Open Event Team
<br/>--
<br/>Login to manage your orders at http://eventyay.com
```

iamareebjamal commented:
That's in English

iamareebjamal commented:
Build is still feeling. Seems like you didn't check the failing build I mentioned

codedsun commented:
> Build is still feeling. Seems like you didn't check the failing build I mentioned

I checked, i thought it would be fixed but still failing. Let me fix it again

codedsun commented:
@iamareebjamal pytype ci also needs to be fixed?

iamareebjamal commented:
Yes

codedsun commented:
@iamareebjamal - need your help for pytype issue, i am unable to resolve this. i am not getting what's the issue.

iamareebjamal commented:
Disable https://google.github.io/pytype/errors.html#mro-error on `app/instance.py", line 243, in get_locale`

iamareebjamal commented:
Please also write the command to extract to pot, merge to po, and generating mo files because I'll need that for testing

codedsun commented:
Extracting to pot
`pybabel extract -F babel.cfg -k _l -o app/translations/messages.pot .`

Updating po
` pybabel update -i app/translations/messages.pot -d app/translations`

Generating mo
`pybabel compile -d app/translations`

codedsun commented:
@iamareebjamal - Done!

iamareebjamal commented:
How did you test the translation was happening?

codedsun commented:
> How did you test the translation was happening?

As you sugggested, In the `de` folder I changed the msgStr of Ticket purchased string `Thank you for your ticket order for` to `Suneet Srivastava` 

i changed my browser display language in firefox to German (De), the website was translated to german now. 
I logged in and purchased a ticket. Now in the mails table I gave you the exact text of the message here. 
https://github.com/fossasia/open-event-server/pull/7416#issuecomment-723538229

iamareebjamal commented:
OK. Thank you

iamareebjamal commented on app/templates/email/ticket_purchased_attendee.html:
```diff
@@ -1,17 +1,17 @@
-    <br/>Quantity: {{ attendee | length }}
+{% for i, attendee in attendees | groupby('ticket_id') %}
+    <br/>{{ _('Ticket') }}: {{ attendee.0.ticket.name }}
+    <br/>{ _('Quantity') }}: {{ attendee | length }}
```
{ missing
